tardis/energy_input/gamma_ray_packet_source.py

```diff
@@ -0,0 +1,856 @@
+        z = np.random.random(no_of_packets)
+        decay_times = z * start + (1 - z) * end
+        return decay_times
+
+    def create_packet_times_uniform_energy(
+        self, no_of_packets, isotopes, decay_time
+    ):
+        """Samples the decay time from the mean lifetime of the isotopes
+
+        Parameters
+        ----------
+        no_of_packets : int
+            Number of packets
+        isotopes : pd.Series
+            Series of packet parent isotopes
+        decay_time : array
+            Series of packet decay time index
+
+        Returns
+        -------
+        array
+            Array of decay times
+        """
+        decay_times = np.zeros(len(no_of_packets))
+        for i, isotope in enumerate(isotopes.to_numpy()):
+            decay_time_min = self.times[decay_time[i]]
+            if decay_time_min == self.times[-1]:
+                decay_time_max = self.effective_times[-1]
+            else:
+                decay_time_max = self.times[decay_time[i] + 1]
+            # rejection sampling
+            while (decay_times[i] <= decay_time_min) or (
+                decay_times[i] >= decay_time_max
+            ):
+                decay_times[i] = -self.taus[isotope] * np.log(
+                    np.random.random()
+                )
+        return decay_times
+
+    def create_packets(
+        self, decays_per_isotope, number_of_packets, *args, **kwargs
+    ):
+        """Initialize a collection of GXPacket objects for the simulation
+        to operate on.
+
+        Parameters
+        ----------
+        decays_per_isotope : array int64
+            Probability of decays per simulation shell per isotope per time step
+        number_of_packets : int
+            Number of packets to create
+
+        Returns
+        -------
+        GXPacketCollection
+        """
+        # initialize arrays for most packet properties
+        locations = np.zeros((3, number_of_packets))
+        directions = np.zeros((3, number_of_packets))
+        packet_energies_rf = np.zeros(number_of_packets)
+        packet_energies_cmf = np.zeros(number_of_packets)
+        nus_rf = np.zeros(number_of_packets)
+        nus_cmf = np.zeros(number_of_packets)
+        times = np.zeros(number_of_packets)
+        statuses = np.ones(number_of_packets, dtype=np.int64) * 3
+
+        self.energy_plot_positron_rows = np.zeros((number_of_packets, 4))
+
+        # compute positronium continuum
+        positronium_energy, positronium_intensity = positronium_continuum()
+
+        # sample packets from dataframe, returning a dataframe where each row is
+        # a sampled packet
+        sampled_packets_df = decays_per_isotope.sample(
+            n=number_of_packets,
+            weights="decay_energy_erg",
+            replace=True,
+            random_state=np.random.RandomState(self.base_seed),
+        )
+        # get unique isotopes that have produced packets
+        isotopes = pd.unique(sampled_packets_df["isotopes"])
```

Comment:
`isotopes` is an index in the data frame.

Reply:
At which level?

Reply:
`time`, `shell_number`, `isotope`